### DISCUSSION TITLE
Scripts in IPXe

### DISCUSSION THREAD

**Discussion by madurani** [ipxe/ipxe#1116] [Q&A]
I am trying unattended installation of windows via IPXe. For it is needed modify some strings in Autounattend.xml file. I thought, that i can use for this purpose apache-cgi configuration. I created python script which can be start via URL with arguments:
http://192.168.100.1/win11pe/win.py?hostname=tml2035&profil=fyzika&file=Autounattend01.xml
mentioned script serve for change of hostname and salt state filename in file Autounattend01.xml. When i past this link to web browser manually, then script success change in mentioned file hostname and name of salt state file.

I tried call this script from IPXE menu but it not work:
```
:wimboot06
set file Autounattend01.xml
chain http://192.168.100.1/win11pe/win.py?hostname=&{cmhostname}&profil=${s_profil}&file=${file} ||
kernel http://${server_ip}/win11pe/wimboot
initrd http://${server_ip}/win11pe/install04.bat        install.bat
initrd http://${server_ip}/win11pe/winpeshl.ini         winpeshl.ini
initrd http://${server_ip}/win11pe/BCD10                BCD
initrd http://${server_ip}/win11pe/boot10.sdi           boot.sdi
initrd http://${server_ip}/win11pe/boot10.wim           boot.wim
boot
```
Could be very suitable, when i can add installations arguments direct in ipxe boot menu and not start manually some scripts before installation on some other PC:
![20240105_132432](https://github.com/ipxe/ipxe/assets/49711883/75aa4718-37af-47d1-9cc9-50188c42fcd3)

Is there some possibility call/start this script properly?
![20240105_130323](https://github.com/ipxe/ipxe/assets/49711883/4a42956d-7182-4d54-940f-a433e73b122e)

**Comment by NiKiZe** [comment]
Read the message and linked page of your "Exec format error" message.
Make sure that it is a valid iPXE script with "#!iPXE" at the first line, be carefull with BOM markers etc.

And a tip if you are using latest version of wimboot, drop your BCD and boot.sdi files, they are not needed.

**Reply by madurani** [reply]
It is python script, it have to begin with #!/usr/bin/python3. I haven't idea how should be look like it. I can't find any how to for CGI-ipxe scripts.

**Reply by NiKiZe** [reply]
Your python script needs to execute on the server and generate a valid iPXE script as it's output.
If you do not get a iPXE script in your browser when testing then you need to resolve that on the server.

How you make your http server run the python code is probably well documented elsewhere. Try searching for "run python script on http server"

**Reply by madurani** [reply]
here is screen , when i manually pass link to web browser:

![image](https://github.com/ipxe/ipxe/assets/49711883/b53dd795-a14d-4ba6-ae6d-bc82642dd500)

**Reply by madurani** [reply]
Python script is started in apache via CGI module:
```
<Directory /volume01/http/win11pe>
    Options +ExecCGI
    AddHandler cgi-script .cgi .pl .py .rb
</Directory>
```

**Reply by madurani** [reply]
here is python script:
```
#!/usr/bin/python3

import os
import sys
import urllib.parse
import re

print('Content-Type: text/plain')
print('')
server_ip = '192.168.1.1'

query_string = os.environ['QUERY_STRING']

arguments = urllib.parse.parse_qs(query_string)

fin = open(str(arguments["file"][0]), "rt")
data = fin.read()
Old_hostname = re.findall("tml[0-9]{4}", data)[0]
Salt_profil = re.findall('informatika|student|katedra|fyzika|default', data)[0]
data = data.replace(Old_hostname, str(arguments["hostname"][0]))
data = data.replace(Salt_profil, str(arguments["profil"][0]))
fin.close()
fin = open(str(arguments["file"][0]), "wt")
fin.write(data)
fin.close()


print ('#!ipxe\n\n')
print ('kernel http://' + server_ip + '/win11pe/wimboot')
print ('initrd http://' + server_ip + '/win11pe/install04.bat        install.bat')
print ('initrd http://' + server_ip + '/win11pe/winpeshl.ini         winpeshl.ini')
print ('initrd http://' + server_ip + '/win11pe/BCD10                BCD')
print ('initrd http://' + server_ip + '/win11pe/boot10.sdi           boot.sdi')
print ('initrd http://' + server_ip + '/win11pe/boot10.wim           boot.wim')
print ('boot')
```

**Reply by NiKiZe** [reply]
Thanks for posting the script

So your script then starts with an empty line
And #!iPXE then comes at the second line!

Another tip, iPXE is capable of relative URLs, you don't need the server address on those initrd lines.

**Comment by madurani** [comment]
Problem was found, syntax error of URL in ipx boot menu :
chain http://192.168.100.1/win11pe/win.py?hostname=&{cmhostname}&profil=${s_profil}&file=${file}
& before {cmhostname}
after change to to $ it work properly:
Change content of Autounattend04.xml file and then start boot.

**Comment by epacua** [comment]
I use plain ipxe scripting when auto-installing Windows 10.  The parameter `--name` is required, without that the files are not injected. I tested that by dropping at WinPE and checking the directory listing.

```
#!ipxe

kernel http://${server}/winpe/wimboot gui
initrd http://${server}/winpe/${arch}/winpe10/boot/fonts/wgl4_boot.ttf wgl4_boot.ttf
initrd http://${server}/winpe/${arch}/winpe10/boot/fonts/segmono_boot.ttf segmono_boot.ttf
initrd http://${server}/winpe/${arch}/winpe10/boot/fonts/segoe_slboot.ttf segoe_slboot.ttf

initrd --name winpeshl.ini http://${server}/winpe/${arch}/startup-scripts/winpe10/winpeshl.ini  winpeshl.ini
initrd --name install.cmd http://${server}/winpe/${arch}/startup-scripts/winpe10/install.cmd    install.cmd
initrd http://${server}/winpe/${arch}/winpe10/bootmgr            bootmgr
initrd http://${server}/winpe/${arch}/winpe10/boot/bcd           bcd
initrd http://${server}/winpe/${arch}/winpe10/boot/boot.sdi      boot.sdi
initrd http://${server}/winpe/${arch}/winpe10/sources/boot.wim   boot.wim

# `boot' is IMPORTANT!!!
boot
```

Contents of winpeshl.ini and install.cmd:
```
[root@san x86_64]# cat winpeshl.ini 
[LaunchApps]
%SYSTEMROOT%\System32\cmd.exe, "/C startnet.cmd"
%SYSTEMROOT%\System32\install.cmd

[root@san x86_64]# cat install.cmd 
net use z: \\san.internal.pacua.net\os-installers /u:op password
cd Z:\x86_64\win10\
Z:
setup.exe /unattend:autounattend.xml

```
Hope this helps.

@NiKiZe , I'm on wimboot-2.7.6 but the lines boot.sdi and bcd are still required, else it fails.

**Reply by NiKiZe** [reply]
boot.sdi and BCD is extracted from .wim
Same for fonts and bootmgr. If not working the .wim might be missing these.

--name is only used to change the name from the original used at the server, check with imgstat.

See https://github.com/NiKiZe/wimboot-install/blob/main/win10.ipxe
Take win10 iso, grab boot.wim and try boot only that without any initrds for a baseline.

**Reply by madurani** [reply]
Is there some possibility interrupted ipxe boot script and not continue with boot system from disk?
In python part of script i have conditions which find out, that in file where are stored installation's windows-keys isn't any free(file is empty).
In this case script render to ipxe condition:
```
    print ('echo Store of keys ' + subor_k + ' is empty')
    print ('echo -n Can be use generic key? (y/n): ${}')
    print ('read status_key')
    print ('iseq ${status_key} y ||')
```
When i chose y, then python set up generic key and continue in boot/installation of new win properly.

How can i define some ipxe condition for interrupted of ipxe script and no continue in boot from disk?
Something as:
```
    print ('iseq ${status_key} n  ?????????
```

**Reply by NiKiZe** [reply]
You already have

     iseq ${status_key} y ||

Make that

    iseq ${status_key} y || goto notwin

And then

    :notwin
    echo not win
    exit

Of course just an example.

**Reply by epacua** [reply]
I forgot to mention that my boot.wim is derived from ADK and not from installer iso.

I'll try the installer boot.wim. Thanks

**Comment by madurani** [comment]
I have again problem. Script worked properly for MBR/UEFI and WIN11/10, but now work only for MBR win11/10. UEFI ipxe boot stop work and i don't know way.

```
kernel http://192.168.1.1/win11pe/wimboot
initrd http://192.168.1.1/win11pe/install01.bat        install.bat
initrd http://192.168.1.1/win11pe/winpeshl.ini         winpeshl.ini
initrd http://192.168.1.1/win11pe/BCD11                BCD
initrd http://192.168.1.1/win11pe/boot11.sdi           boot.sdi
initrd http://192.168.1.1/win11pe/boot11.wim           boot.wim
boot
```
![image](https://github.com/ipxe/ipxe/assets/49711883/8e4f2daa-6671-4615-b3b0-4277dda4488c)

![image](https://github.com/ipxe/ipxe/assets/49711883/e38e81fb-66c7-4f43-810c-6da644168fd9)

**Reply by NiKiZe** [reply]
As @epacua mentioned above --name is required for EFI, if the name you want differs for the name in url.

See examples above or https://github.com/NiKiZe/wimboot-install/blob/main/win10.ipxe where files with different names are renamed. While those that don't need a rename is not.
Also consider using orginal boot.wim but cleaned using methods in that repo, instead of ADK, it will most likely make your life easier.

**Comment by madurani** [comment]
Not help, BCD problem persist. 
```
initrd --name install.bat http://192.168.1.1/win11pe/install01.bat        install.bat
initrd --name winpeshl.ini http://192.168.1.1/win11pe/winpeshl.ini         winpeshl.ini
initrd --name BCD http://192.168.1.1/win11pe/BCD11                BCD
initrd --name boot.sdi http://192.168.1.1/win11pe/boot11.sdi           boot.sdi
initrd --name boot.wim http://192.168.1.1/win11pe/boot11.wim           boot.wim

```

**Reply by madurani** [reply]
What need option --name ? MBR, UEFI or version of wine 10/11.
Now began work, but I need exactly to know what need option --name and what no.

**Reply by NiKiZe** [reply]
--name is required in EFI mode for any file that should have a different name in wimboot than it has on the server.

In the example I pointed to you can see that name is only given if it differs from server.

Use imgstat command to verify what you have befor boot.

**Comment by madurani** [comment]
Thanks for your help, i added --name options for all win profiles and it look like that work fine. I did tests:
UEFI Win11 >>> ok
UEFI Win10 >>> ok
MBR Win11 >>> ok
MBR Win10 >>> ok
